tests/daemon/test-circular-file.c

```diff
@@ -609,11 +611,158 @@ test_circular_file_shrink (Fixture      *fixture,
   g_object_unref (circular_file_2);
 }
 
+static void
+test_circular_file_broken_metadata_file (Fixture      *fixture,
+                                         const gchar  *metadata_file_contents,
+                                         gssize        metadata_file_length,
+                                         GKeyFileError expected_error_code)
+{
+  const gsize max_size = 1024;
+  g_autofree gchar *data_file_contents = g_malloc (max_size);
+  g_autofree gchar *metadata_file_path =
+    g_strconcat (fixture->data_file_path, METADATA_EXTENSION, NULL);
+  g_autoptr(GError) error = NULL;
+  gboolean ret;
+
+  /* Fill the data file with junk. */
+  memset (data_file_contents, '*', max_size);
+  ret = g_file_set_contents (fixture->data_file_path, data_file_contents, max_size, &error);
+  g_assert_no_error (error);
+  g_assert_true (ret);
+
+  /* Initialize the metadata file to the desired incorrect contents */
+  ret = g_file_set_contents (metadata_file_path, metadata_file_contents, metadata_file_length, &error);
+  g_assert_no_error (error);
+  g_assert_true (ret);
+
+  g_autoptr(EmerCircularFile) circular_file = NULL;
+
+  circular_file = emer_circular_file_new (fixture->data_file_path, max_size,
+                                          /* reinitialize */ FALSE,
+                                          &error);
+  g_assert_null (circular_file);
+  g_assert_error (error, G_KEY_FILE_ERROR, expected_error_code);
+  g_clear_error (&error);
+
+  circular_file = emer_circular_file_new (fixture->data_file_path, max_size,
+                                          /* reinitialize */ TRUE,
+                                          &error);
+  g_assert_no_error (error);
+  g_assert_nonnull (circular_file);
+
+  /* Verify that the file is logically empty, even though it's physically full
+   * of asterisks.
+   */
+  const gchar * const NO_STRINGS[0];
+  guint64 token = assert_strings_read (circular_file, NO_STRINGS, 0);
+  g_assert_false (emer_circular_file_has_more (circular_file, token));
+
+  /* Adding new stuff to the file should work. */
+  const gchar * const STRINGS[] =
+    {
+      "Sneezy", "phylum", "Europe", "sloth", "guacamole", "data link",
+      "Colossus of Rhodes"
+    };
+  gsize NUM_STRINGS = G_N_ELEMENTS (STRINGS);
+  assert_strings_saved (circular_file, STRINGS, NUM_STRINGS);
+}
+
+static void
+test_circular_file_metadata_file_empty (Fixture      *fixture,
+                                        gconstpointer unused)
+{
+  /* If the metadata file exists but is empty, we should initialize it and
+   * consider the circular file itself to be empty.
+   */
+  test_circular_file_broken_metadata_file (fixture, "", 0,
+                                           G_KEY_FILE_ERROR_GROUP_NOT_FOUND);
```

Comment:
You'll have to update the expectation here if you finally silently deal with empty files and add the missing group/keys there, I think.